src/main/kotlin/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtSharedTableBackupManager.kt

```diff
@@ -0,0 +1,456 @@
+                    ListObjectsV2Request()
+                            .withBucketName(s3BucketName)
+                            .withContinuationToken(continuationToken)
+                            .withPrefix("$backupDir/$backupId/${tenantTable.tenantName}/${tenantTable.virtualTableName}"))
+            continuationToken = listBucketResult.continuationToken
+            for (o in listBucketResult.objectSummaries) {
+                ret.add(o.key)
+            }
+        } while (listBucketResult.isTruncated)
+        return ret
+    }
+
+    override fun listBackups(listBackupRequest: ListBackupsRequest): ListBackupsResult {
+        val ret = Lists.newArrayList<BackupSummary>()
+
+        Preconditions.checkArgument(listBackupRequest.backupType == null, "Listing backups by backupType unsupported")
+        Preconditions.checkArgument(listBackupRequest.tableName == null, "Listing backups by table name unsupported for multi tenant backups")
+        Preconditions.checkArgument(listBackupRequest.timeRangeLowerBound == null, "Listing backups filtered by time range unsupported [currently]")
+        Preconditions.checkArgument(listBackupRequest.timeRangeUpperBound == null, "Listing backups filtered by time range unsupported [currently]")
+
+        // translate the last backupArn we saw to the file name on S3, to iterate from
+        val startAfter: String? = if (listBackupRequest.exclusiveStartBackupArn == null) null
+            else getBackupMetadataFile(listBackupRequest.exclusiveStartBackupArn)
+
+        val listBucketResult: ListObjectsV2Result = s3.listObjectsV2(
+                ListObjectsV2Request()
+                        .withMaxKeys(listBackupRequest.limit)
+                        .withDelimiter("/")
+                        .withBucketName(s3BucketName)
+                        .withStartAfter(startAfter)
+                        .withPrefix(backupMetadataDir + "/"))
+        for (o in listBucketResult.objectSummaries) {
+            val backup = mtBackupAwsAdaptor.getBackupSummary(getBackup(getBackupIdFromKey(o.key))!!)
+            ret.add(backup)
+        }
+        var lastEvaluatedBackupArn: String? = null
+        if (listBucketResult.isTruncated) {
+            lastEvaluatedBackupArn = Iterables.getLast(ret).backupArn
+        }
+        return ListBackupsResult().withBackupSummaries(ret)
+                .withLastEvaluatedBackupArn(lastEvaluatedBackupArn)
+    }
+
+    override fun getBackup(id: String): MtBackupMetadata? {
+        try {
+            val backupFile: S3Object = s3.getObject(s3BucketName, getBackupMetadataFile(id))
+            val backupFileString = CharStreams.toString(
+                    InputStreamReader(backupFile.objectContent.delegateStream, charset))
+            return gson.fromJson(backupFileString,
+                    MtBackupMetadata::class.java)
+        } catch (e: AmazonS3Exception) {
+            if ("NoSuchKey".equals(e.errorCode)) {
```

Comment:
Note: you may get this even if the object exists: https://codeburst.io/quick-explanation-of-the-s3-consistency-model-6c9f325e3f82

Reply:
I think we can live with that risk till we move metadata to dynamo, I mostly just want to see some real backups and restores in action before spending time on that